tower-timeout/src/lib.rs

```diff
@@ -83,50 +98,114 @@ where T: Future,
         match self.response.poll() {
             Ok(Async::Ready(v)) => return Ok(Async::Ready(v)),
             Ok(Async::NotReady) => {}
-            Err(e) => return Err(Error::Inner(e)),
+            Err(e) => return Err(Error(Kind::Inner(e))),
         }
 
         // Now check the sleep
         match self.sleep.poll() {
             Ok(Async::NotReady) => Ok(Async::NotReady),
-            Ok(Async::Ready(_)) => Err(Error::Timeout),
-            Err(_) => Err(Error::Timeout),
+            Ok(Async::Ready(_)) => Err(Error(Kind::Elapsed)),
+            Err(_) => Err(Error(Kind::Elapsed)),
         }
     }
 }
 
-
 // ===== impl Error =====
 
 impl<T> fmt::Display for Error<T>
 where
     T: fmt::Display,
 {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match *self {
-            Error::Inner(ref why) => fmt::Display::fmt(why, f),
-            Error::Timeout => f.pad("request timed out"),
+        match self.0 {
+            Kind::Inner(ref why) => fmt::Display::fmt(why, f),
+            Kind::Elapsed => f.pad("request timed out"),
+            Kind::Timer(ref why) => fmt::Display::fmt(why, f),
         }
     }
 }
 
 impl<T> error::Error for Error<T>
 where
-    T: error::Error,
+    T: error::Error + 'static,
 {
-    fn cause(&self) -> Option<&error::Error> {
-        if let Error::Inner(ref why) = *self {
+    fn description(&self) -> &str {
+        match self.0 {
+            Kind::Inner(_) => "inner service error",
```

Comment:
Since `T` is known to implement `Error`, this should probably be
```suggestion
            Kind::Inner(e) => e.description(),
```